Mathlib/Analysis/LocallyConvex/AbsConvex.lean

```diff
@@ -70,8 +161,201 @@ theorem nhds_basis_abs_convex_open :
+        apply congrArg (HSMul.hSMul (1 / 2)) h
+      simp only [one_div, ne_eq, OfNat.ofNat_ne_zero, not_false_eq_true, inv_smul_smul₀] at e1
+      exact e1
+    have e11 : x-(1 / 2 : ℝ) • (x + y)  = z' := by
+      rw [smul_add]
+      apply help
+      rw [smul_sub, smul_add, ← smul_assoc]
+      simp only [one_div, smul_eq_mul, isUnit_iff_ne_zero, ne_eq, OfNat.ofNat_ne_zero,
+        not_false_eq_true, IsUnit.mul_inv_cancel, one_smul, smul_inv_smul₀, smul_neg]
+      rw [add_comm]
+      rw [two_smul]
+      simp only [add_sub_add_right_eq_sub]
+      rw [← hz₂]
+      simp only [add_sub_cancel_left]
+      exact id (Eq.symm hz'₂)
+    have e12 : (1 / 2 : ℝ) • (x + y) - y  = z' := by
+      apply help
+      rw [smul_sub]
+      simp only [one_div, smul_add, ne_eq, OfNat.ofNat_ne_zero, not_false_eq_true, smul_inv_smul₀,
+        smul_neg]
+      rw [two_smul, add_sub_add_right_eq_sub, ← hz₂]
+      rw [add_sub_cancel_left]
+      exact id (Eq.symm hz'₂)
+    constructor
+    · rw [mem_preimage, e12]
+      exact hz'₁
+    · rw [mem_preimage, e11]
+      exact hz'₁
+
+lemma balancedHull_subset_convexHull_union_neg {s : Set E} :
+    balancedHull ℝ s ⊆ convexHull ℝ (s ∪ -s) := by
+  intro a ha
+  obtain ⟨r, hr, y, hy, rfl⟩ := mem_balancedHull_iff.1 ha
+  apply segment_subset_convexHull (mem_union_left (-s) hy) (mem_union_right _ (neg_mem_neg.mpr hy))
+  use (1+r)/2
+  use (1-r)/2
+  constructor
+  · rw [← zero_div 2]
+    exact (div_le_div_right zero_lt_two).mpr (neg_le_iff_add_nonneg'.mp (neg_le_of_abs_le hr))
+  · constructor
+    · rw [← zero_div 2]
+      exact (div_le_div_right zero_lt_two).mpr (sub_nonneg_of_le (le_of_max_le_left hr))
+    · constructor
+      · ring_nf
+      · rw [smul_neg, ← sub_eq_add_neg, ← sub_smul]
+        apply congrFun (congrArg HSMul.hSMul _) y
+        ring_nf
+
+@[simp]
+theorem convexHull_union_neg_eq_absConvexHull {s : Set E} :
+    convexHull ℝ (s ∪ -s) = absConvexHull ℝ s := by
+  rw [absConvexHull_eq_convexHull_balancedHull]
+  exact le_antisymm (convexHull_mono (union_subset (subset_balancedHull ℝ)
+    (fun _ _ => by rw [mem_balancedHull_iff]; use -1; aesop)))
+    (by
+      rw [← Convex.convexHull_eq (convex_convexHull ℝ (s ∪ -s))]
+      exact convexHull_mono balancedHull_subset_convexHull_union_neg)
+
+
+lemma add_self_eq_smul_two {V : Set E} (h : Convex ℝ V) : V + V = (2 : ℝ) • V := by
+  rw [← one_add_one_eq_two, Convex.add_smul h (zero_le_one' ℝ) (zero_le_one' ℝ), MulAction.one_smul]
+
+variable (E 𝕜) {s : Set E}
+variable [NontriviallyNormedField 𝕜]  [Module 𝕜 E]
+variable [SMulCommClass ℝ 𝕜 E]
+variable [TopologicalSpace E] [TopologicalAddGroup E]  [lcs : LocallyConvexSpace ℝ E]
+  [ContinuousSMul ℝ E]
+
+-- TVS II.25 Prop3
+theorem totallyBounded_absConvexHull
+    (hs : TotallyBounded (uniformSpace := TopologicalAddGroup.toUniformSpace E) s) :
+    TotallyBounded (uniformSpace := TopologicalAddGroup.toUniformSpace E) (absConvexHull ℝ s) := by
```

Comment:
The better way to state this would be to assume `[UniformSpace E] [UniformAddGroup E]`. It feels a bit weird from the math point of view since the uniformity comes from the topology, but in many case it won't be defeq to `TopologicalAddGroup.toUniformSpace` (e.g your current lemma does not apply to normed spaces !). And it avoids a bunch of troubles in the proof.

Reply:
Another thing : I think I would rather first state the result about the convex hull, and then use `convexHull_union_neg_eq_absConvexHull` to deduce the current result. I know Bourbaki does it the other way around (after all, what you wrote is stronger), but since we will have to *state* the convex hull statement anyway (which they don't !) I think we might as well split up the proof.
But of course this is a matter of taste

Reply:
> The better way to state this would be to assume `[UniformSpace E] [UniformAddGroup E]`

@ADedecker Do I also need a hypothesis `(h : U = TopologicalAddGroup.toUniformSpace E)` to connect the uniformity to the locally convex topology?

Reply:
Ah, I can use `UniformAddGroup.toUniformSpace_eq`

Reply:
Sorry for the delay. Yes, you can use that, but it sounds a bit weird that you have to use it, let me have a look.

Reply:
Well indeed you don't use it in the current version!

Reply:
@ADedecker yes - originally I added `[UniformSpace E]` and `[UniformAddGroup E]` without removing `[TopologicalSpace E]` and `[TopologicalAddGroup E]`. Removing `[TopologicalSpace E]` and `[TopologicalAddGroup E]` resolved the issue.

Reply:
> I think I would rather first state the result about the convex hull, and then use `convexHull_union_neg_eq_absConvexHull` to deduce the current result.

I have done this now.